crates/bevy_app/src/app_builder.rs

```diff
@@ -235,8 +235,10 @@ impl AppBuilder {
     where
         R: FromResources + Send + Sync + 'static,
     {
-        let resource = R::from_resources(&self.app.resources);
-        self.app.resources.insert(resource);
+        if !self.resources().contains::<R>() {
```

Comment:
This does "double hash" the type id. Resources don't currently have an "entry" api. And I don't think adding one is in scope for this pr. But can we add a `// PERF: remove this double hash` comment so it is discoverable?

Reply:
That would be difficult in this case, because we need to both be modifying `self.resources()` for the `Entry`, and reading it for the `from_resources` call - so without some gnarly unsafe code, I'm not sure this is feasible. 
It's also worth noting that this is using the `bevy_utils::HashMap`, so the hashing is rather cheap.
Also this only happens at app startup time, so the cost doesn't matter too much. I've added a comment explaining this though

Reply:
makes sense. i added a "PERF" label here. I use those like TODO labels, but for places where we are knowingly leaving perf on the table.